Don't desugar `let _ = local_ x`

Julow opened the pull request:
The desugared syntax `let local_ _ = x` is invalid.

Same as https://github.com/janestreet/ocamlformat/pull/20, rebased on the `jane` branch.

Julow committed:
Don't desugar `let _ = local_ x`

The desugared syntax `let local_ _ = x` is invalid.

Signed-off-by: Jules Aguillon <jules@j3s.fr>

Julow committed:
Restrict `let local_ ...` rewrite even more

This restricts the desugaring of `let x = local_ y` to the cases where
`x` is an identifier or an identifier with a type annotation.

Signed-off-by: Jules Aguillon <jules@j3s.fr>

Julow commented:
I noticed more cases of invalid syntax and restricted the desugaring even more.

Examples of invalid syntaxes that were generated:
```ocaml
let local_ () = x
let local_ {b} = y
```

ccasin commented on lib/Sugar.ml:
```diff
@@ -326,8 +326,17 @@ module Let_binding = struct
+               annotations and extensions into the pattern and the
+               expression. *)
+            match lb_pat.ppat_desc with
+            | Ppat_var _ | Ppat_constraint ({ppat_desc = Ppat_var _;_}, _) ->
```
If I understand correctly, this will cause `let (x : string) = local_ "hi" in ...` to be rewritten to `let local_ (x : string) = "hi" in ...`.  But `local_` only works on a let binding if the pattern is exactly a variable.

Julow replied:
This indeed cause a syntax error but an equivalent syntax exists:

```ocaml
let local_ x : string = local_ "hi"
```

This rewrite could be allowed by improving the formatting function.

Julow replied:
The formatting function is not the only problem. The AST of these two lines is different:
```ocaml
let local_ x : string = local_ "hi"
let (x : string) = local_ "hi"
```

I added a commit to distinguish between these two cases (and a third one more complex). The problem is fixed but the code is complex and the rewrite don't happen when it could.
To go further, the parser should be changed to place the constraint in the same place in both cases.

Here are the ASTs:

<details>

Line 1:

```
Ptop_def
  [
    structure_item (_none_[0,0+-1]..[0,0+-1]) ghost
      Pstr_value Nonrec
      [
        <def>
          pattern (_none_[0,0+-1]..[0,0+-1]) ghost
            Ppat_var "x" (_none_[0,0+-1]..[0,0+-1]) ghost
          expression (_none_[0,0+-1]..[0,0+-1]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local"
              []
            [
              <arg>
              Nolabel
                expression (_none_[0,0+-1]..[0,0+-1]) ghost
                  attribute "extension.local"
                    []
                  Pexp_constraint
                  expression (_none_[0,0+-1]..[0,0+-1]) ghost
                    Pexp_constant PConst_string("hi",(_none_[0,0+-1]..[0,0+-1]) ghost,None)
                  core_type (_none_[0,0+-1]..[0,0+-1]) ghost
                    Ptyp_constr "string" (_none_[0,0+-1]..[0,0+-1]) ghost
                    []
            ]
      ]
  ]
```

Line 2:

```
Ptop_def
  [
    structure_item (_none_[0,0+-1]..[0,0+-1]) ghost
      Pstr_value Nonrec
      [
        <def>
          pattern (_none_[0,0+-1]..[0,0+-1]) ghost
            Ppat_constraint
            pattern (_none_[0,0+-1]..[0,0+-1]) ghost
              Ppat_var "x" (_none_[0,0+-1]..[0,0+-1]) ghost
            core_type (_none_[0,0+-1]..[0,0+-1]) ghost
              Ptyp_constr "string" (_none_[0,0+-1]..[0,0+-1]) ghost
              []
          expression (_none_[0,0+-1]..[0,0+-1]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local"
              []
            [
              <arg>
              Nolabel
                expression (_none_[0,0+-1]..[0,0+-1]) ghost
                  Pexp_constant PConst_string("hi",(_none_[0,0+-1]..[0,0+-1]) ghost,None)
            ]
      ]
  ]
```

</details>

ccasin replied:
I am surprised by the first AST here.  The parser in the compiler includes a `Ppat_constraint` in the pattern for this example, and I am having trouble finding where the ocamlformat parsers differ in a way that would account for the absence of this node.

Julow replied:
I don't know what I did wrong in my previous comment but the ASTs are:

line 1:

```
Ptop_def
  [
    structure_item (./test.ml[1,0+0]..[1,0+35])
      Pstr_value Nonrec
      [
        <def> (./test.ml[1,0+0]..[1,0+35])
          pattern (./test.ml[1,0+11]..[1,0+21]) ghost
            attribute "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost (_none_[0,0+-1]..[0,0+-1]) ghost
              []
            Ppat_constraint
            pattern (./test.ml[1,0+11]..[1,0+12])
              Ppat_var "x" (./test.ml[1,0+11]..[1,0+12])
            core_type (./test.ml[1,0+15]..[1,0+21]) ghost
              Ptyp_poly
              core_type (./test.ml[1,0+15]..[1,0+21])
                Ptyp_constr "string" (./test.ml[1,0+15]..[1,0+21])
                []
          expression (./test.ml[1,0+4]..[1,0+35]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost
              []
            [
              <arg>
              Nolabel
                expression (./test.ml[1,0+4]..[1,0+35])
                  attribute "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost (_none_[0,0+-1]..[0,0+-1]) 
ghost
                    []
                  Pexp_constraint
                  expression (./test.ml[1,0+24]..[1,0+35]) ghost
                    Pexp_apply
                    expression (_none_[0,0+-1]..[0,0+-1]) ghost
                      Pexp_extension "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost
                      []
                    [
                      <arg>
                      Nolabel
                        expression (./test.ml[1,0+31]..[1,0+35])
                          Pexp_constant
                          constant (./test.ml[1,0+31]..[1,0+35])
                            PConst_string("hi",(./test.ml[1,0+32]..[1,0+34]),None)
                    ]
                  core_type (./test.ml[1,0+15]..[1,0+21])
                    Ptyp_constr "string" (./test.ml[1,0+15]..[1,0+21])
                    []
            ]
      ]
  ]
```

line 2:

```
Ptop_def
  [
    structure_item (./test.ml[2,36+0]..[2,36+30])
      Pstr_value Nonrec
      [
        <def> (./test.ml[2,36+0]..[2,36+30])
          pattern (./test.ml[2,36+4]..[2,36+16])
            Ppat_constraint
            pattern (./test.ml[2,36+5]..[2,36+6])
              Ppat_var "x" (./test.ml[2,36+5]..[2,36+6])
            core_type (./test.ml[2,36+9]..[2,36+15])
              Ptyp_constr "string" (./test.ml[2,36+9]..[2,36+15])
              []
          expression (./test.ml[2,36+19]..[2,36+30]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost
              []
            [
              <arg>
              Nolabel
                expression (./test.ml[2,36+26]..[2,36+30])
                  Pexp_constant
                  constant (./test.ml[2,36+26]..[2,36+30])
                    PConst_string("hi",(./test.ml[2,36+27]..[2,36+29]),None)
            ]
      ]
  ]
```

Both cases have a `Ppat_constraint` indeed. One have the constraint on the expression, while the other doesn't.

Julow committed:
Tests: Add a .ref file for local.ml

There are syntax rewrite on locals, it's important not to change the
test input.

Julow committed:
Fix local_ desugaring generating invalid syntax

These two items don't have the same AST and must be recognized:

    let local_ x : string = "hi"
    let (x : string) = local_ "hi"

ccasin commented on lib/Sugar.ml:
```diff
@@ -319,15 +319,43 @@ module Let_binding = struct
+        let local_ x : 'a -> 'a -> 'a = "hi" (* Annotation on the pattern *)
+
+        let (x : string) = local_ "hi" (* Annotation on the pattern *)
+      ]} *)
```
I'm confused.  I took a look at the parsetree produced by the compiler for the first two examples here, and they are the same.  So I suppose there is a difference in the ocamlformat parsers - is there an easy way to get ocamlformat to dump the parsetree so I can understand better? (I don't see a -dparsetree equivalent in the options).

Julow replied:
I use `dune exec -- tools/printast/printast.exe ./test.ml` to read the AST.

The two examples indeed have the same AST, I fixed the comment in a new commit.

ccasin replied:
Thanks!

I think an unfortunate related case is:
```ocaml
let x : 'a . 'a -> 'a -> 'a = local_ "hi"
```
If I understand correctly, this will currently be rewritten in such a way as to fail the roundtrip test because we are missing the attribute on the pattern.  I'm not sure what the best resolution is - we could check perhaps for the attribute when deciding to sugar?

Also, this comment is very helpful, but I think it could be slightly clearer in a few ways:
- Because it's a comment involving `local_`, my first guess when reading it was that "annotation on the ..." was a reference to how `local_` is parsed, but I see now it's a reference to how the type annotation is parsed.  So perhaps say "type annotation" or "constraint node" to make it more obvious.
- The comment and the name of the function refer to "sugar" but don't say what is being sugared.  One of them should mention it's a question about when we can sugar "local_" in particular.
- I think in the first case the type annotation ends up on both the expression and the pattern, not just the expression.

ccasin replied:
Sorry, one more annoying request: The comment should probably explain how the location comparison in the code is accomplishing your goal (though I haven't fully understood this bit yet).

Julow replied:
The function name is unclear but it's specifically about "local lets" (whether it can be rewritten into `let local_ x ...`).
I'll improve that and look at the new case.

This code is a hack and not the solution. The solution is to work on the parsers:
- The standard parser should output identical ASTs for pretended-identical syntaxes. This makes the rules of what we can rewrite clearer.
- The extended parser should output different ASTs for the different syntaxes. This avoids the need to guess the current syntax and the possible rewrites.

Julow replied:
As an example, there's `Pexp_record`. The standard parser output the same AST for many syntaxes (`{ x : t = x }` is `{ x = (x : t} }`) but the extended parser output a different AST in every parser rules. (the rules are unchanged, only their output is different)

The `Sugar` function for record expressions have been removed, instead we explicitly rewrite the AST in `Extended_ast.Parse.normalize_mapper`.
The formatting code for records is straightforward.

ccasin replied:
All this makes sense - but what should we do in the short term about failing on
```ocaml
let x : 'a . 'a -> 'a -> 'a = local_ "hi"
```
?

If my suggestion to check for the presence of the attribute on the pattern and not sugar in that case works, perhaps you can just add that to this PR, perhaps add a little more text to the comment to explain the code and document that we intend to come back, and then we can merge this.

If that doesn't work and no other easy fix is available, then we can just merge it as is, with a comment documenting the issue.

Or are you proposing we shouldn't merge this at all and wait until someone has time to change the extended AST?

ccasin replied:
(Also, the CI tests are currently failing on this PR.  I think we expect the windows test to fail, but not the others - probably it just needs to be rebased).

Julow replied:
This PR is already fixing bugs, so it can be merged at any time. I'll look at this case and CI shortly.

Julow replied:
I fixed the bug you mentioned and did a rebase. I hope the comment is clearer too. 

The fix is mostly to not rewrite into the shorter form when that would mess with the formatting code or change the AST. A more concrete AST will allow to remove the tricky code and to rewrite in more cases.

ccasin replied:
Great - thanks for dealing with all these requests!

Julow committed:
Preserve `local_` position when ptyp_poly constraint

These two items don't have the same AST:

    let x : 'a . 'a -> 'a = local_ "hi"
    let local_ x : 'a. 'a -> 'a = "hi"

ccasin reviewed (approved):
Looks good now!
